<h1>Errors: Called on incompatible type</h1> <h2 id="Message">Message</h2> <pre class="syntaxbox">TypeError: 'this' is not a Set object (EdgE)
TypeError: Function.prototype.toString called on incompatible object (Firefox)
TypeError: Function.prototype.bind called on incompatible target (Firefox)
TypeError: Method Set.prototype.add called on incompatible receiver undefined (Chrome)
TypeError: Bind must be called on a function (Chrome)
</pre> <h2 id="Error_type">Error type</h2> <p><a href="../global_objects/typeerror"><code>TypeError</code></a></p> <h2 id="What_went_wrong">What went wrong?</h2> <p>When this error is thrown, a function (on a given object), is called with a <code>this</code> not corresponding to the type expected by the function.</p> <p>This issue can arise when using the <a href="../global_objects/function/call"><code>Function.prototype.call()</code></a> or <a href="../global_objects/function/apply"><code>Function.prototype.apply()</code></a> methods, and providing a <code>this</code> argument which does not have the expected type.</p> <p>This issue can also happen when providing a function (stored in an object) as an argument to another function. In this case, the object won't the <code>this</code> target of the function. To work-around this issue, you will either need to provide a lambda which is making the call, or use the <a href="../global_objects/function/bind"><code>Function.prototype.bind()</code></a> function to force the <code>this</code> argument to the expected object.</p> <h2 id="Examples">Examples</h2> <h3 id="Invalid_cases">Invalid cases</h3> <pre data-language="js">var mySet = new Set;
['bar', 'baz'].forEach(mySet.add);
// mySet.add is a function, but "mySet" is not captured as this.

var myFun = function () {
  console.log(this);
};
['bar', 'baz'].forEach(myFun.bind);
// myFun.bind is a function, but "myFun" is not captured as this.

</pre> <h3 id="Valid_cases">Valid cases</h3> <pre data-language="js">var mySet = new Set;
['bar', 'baz'].forEach(mySet.add.bind(mySet));
// This works due to binding "mySet" as this.

var myFun = function () {
  console.log(this);
};
['bar', 'baz'].forEach(x =&gt; myFun.bind(x));
// This works using the "bind" function. It creates a lambda forwarding the argument.

</pre> <h2 id="See_also">See also</h2> <ul> <li><a href="../global_objects/function/call"><code>Function.prototype.call()</code></a></li> <li><a href="../global_objects/function/apply"><code>Function.prototype.apply()</code></a></li> <li><a href="../global_objects/function/bind"><code>Function.prototype.bind()</code></a></li> </ul><div class="_attribution">
  <p class="_attribution-p">
    <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/Called_on_incompatible_type$edit" class="_attribution-link">Edit this page on MDN</a>
  </p>
</div>
<div class="_attribution">
  <p class="_attribution-p">
    &copy; 2005&ndash;2018 Mozilla Developer Network and individual contributors.<br>Licensed under the Creative Commons Attribution-ShareAlike License v2.5 or later.<br>
    <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/Called_on_incompatible_type" class="_attribution-link">https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/Called_on_incompatible_type</a>
  </p>
</div>
